include/unifex/at_coroutine_exit.hpp

```diff
@@ -0,0 +1,283 @@
+      return unifex::connect(
+          (Sender&&) sender_,
+          _die_on_done_rec_t<Receiver>{(Receiver&&) rec});
+    }
+
+    Sender sender_;
+  };
+};
+
+template <typename Sender>
+using _die_on_done_t =
+    typename _die_on_done<remove_cvref_t<Sender>>::type;
+
+struct _die_on_done_fn {
+  template (typename Value)
+    (requires (!detail::_awaitable<Value>) AND sender<Value>)
+  _die_on_done_t<Value> operator()(Value&& value) /*mutable*/
+      noexcept(std::is_nothrow_constructible_v<remove_cvref_t<Value>, Value>) {
+    return _die_on_done_t<Value>{(Value&&) value};
+  }
+
+  template <typename Value>
+  Value&& operator()(Value&& value) const noexcept {
+    return (Value&&) value;
+  }
+};
+
+template <typename... Ts>
+struct _cleanup_task;
+
+template <typename... Ts>
+struct _cleanup_promise : _cleanup_promise_base {
+  template <typename Action>
+  explicit _cleanup_promise(Action&&, Ts&... ts) noexcept
+    : args_(ts...) {}
+
+  _cleanup_task<Ts...> get_return_object() noexcept {
+    return _cleanup_task<Ts...>(
+        coro::coroutine_handle<_cleanup_promise>::from_promise(*this));
+  }
+
+  coro::coroutine_handle<> unhandled_done() noexcept {
+    // Record that we are processing an unhandled done signal. This is checked in
+    // the final_suspend of the cleanup action to know which subsequent continuation
+    // to resume.
+    isUnhandledDone_ = true;
+    // On unhandled_done, run the cleanup action:
+    return coro::coroutine_handle<_cleanup_promise>::from_promise(*this);
+  }
+
+  template <typename Value>
+  decltype(auto) await_transform(Value&& value)
+      noexcept(noexcept(
+          unifex::await_transform(*this, _die_on_done_fn{}((Value&&) value)))) {
+    return unifex::await_transform(*this, _die_on_done_fn{}((Value&&) value));
+  }
+
+  std::tuple<Ts&...> args_;
+};
+
+template <typename... Ts>
+struct [[nodiscard]] _cleanup_task {
+  using promise_type = _cleanup_promise<Ts...>;
+
+  explicit _cleanup_task(coro::coroutine_handle<promise_type> coro) noexcept
+    : continuation_(coro) {}
+
+  _cleanup_task(_cleanup_task&& that) noexcept
+    : continuation_(std::exchange(that.continuation_, {})) {}
+
+  ~_cleanup_task() {
+    UNIFEX_ASSERT(!continuation_);
+  }
+
+  bool await_ready() const noexcept {
+    return false;
+  }
+
+  template <typename Promise>
+  bool await_suspend(coro::coroutine_handle<Promise> parent) noexcept {
+    continuation_.promise().continuation_ =
+        exchange_continuation(parent.promise(), continuation_);
+    return false;
+  }
+
+  std::tuple<Ts&...> await_resume() noexcept {
+    return std::exchange(continuation_, {}).promise().args_;
+  }
+
+private:
+  continuation_handle<promise_type> continuation_;
+};
+
+namespace _at_coroutine_exit {
+  inline constexpr struct _fn {
+  private:
+    template <typename Action, typename... Ts>
+    static _cleanup_task<Ts...> at_coroutine_exit(Action action, Ts... ts) {
+      co_await std::move(action)(std::move(ts)...);
```

Comment:
Should we be checking that the result of the `co_await` expression is `void` here?
Or are you just relying on `[[nodiscard]]` warnings if values are discarded?

Reply:
This is just my opinion, but it feels overly nannying to enforce that the function returns a void sender.